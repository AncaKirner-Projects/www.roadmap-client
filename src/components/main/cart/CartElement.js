import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';
import styled from 'styled-components';
import ResponsiveAlertDialog from '../ResponsiveAlertDialog';
import ActionTypes from '../../../store/actions/actionTypes';
import { deleteFromCart, changeQuantity } from '../../../store/actions/cartActions';

const PaperStyle = styled(Paper)`
  margin: 5px;
  padding: 15px;
  min-height: 80px;
`;

const PriceStyle = styled(Grid)`
  width: 10%;
`;

class CartElement extends Component {
  state = {
    openAlertDialog: false,
    quantity: this.props.quantity
  }

  handleAlertDialogClose = () => {
    this.setState({ openAlertDialog: false });
  }

  handleRemoveFromCart = (product) => {
    this.props.deleteFromCart(product);
    this.setState({ openAlertDialog: false });
  }

  handleConfirmation = () => {
    this.setState({ openAlertDialog: true });
  }

  handleQuantityChange = product => (e) => {
    this.setState({ quantity: e.target.value });
    this.props.changeQuantity(product, e.target.value);
  }

  render() {
    const { product } = this.props;

    const quantities = () => {
      let res = [];
      for (let i = 1; i <= product.prod_number; i++) {
        res.push({ value: i });
      }

      return res;
    };

    const classes = {
      root: {
        overflow: 'hidden',
        padding: `0 ${10 * 3}px`
      },
      wrapper: {
        maxWidth: 400,
        margin: 100
      },
      paper: {
        margin: 100,
        padding: 100 * 2
      },
      menu: {
        width: 200
      }
    };

    return (
      <PaperStyle className={classes.paper}>
        <Grid container wrap="nowrap" spacing={16}>
          <Grid item xs zeroMinWidth>
            <Typography variant="subtitle1" noWrap>{product.prod_name}</Typography>
          </Grid>
          <Grid item xs zeroMinWidth>
            <TextField
              id="outlined-select-currency-native"
              select
              className={classes.textField}
              onChange={this.handleQuantityChange(product)}
              value={this.state.quantity}
              variant="outlined"
            >
              {
                quantities().map(option => (
                  <option key={option.value} value={option.value}>
                    {option.value}
                  </option>
                ))}
            </TextField>
          </Grid>
          <PriceStyle item zeroMinWidth>
            <Typography align="left" variant="subtitle1" noWrap>{product.price * product.quantity}  &nbsp; RON</Typography>
            <Button
              size="small"
              onClick={this.handleConfirmation}
            >
              Remove
            </Button>
          </PriceStyle>
        </Grid>
        <ResponsiveAlertDialog
          open={this.state.openAlertDialog}
          handleClose={this.handleAlertDialogClose}
          handleConfirmClose={() => this.handleRemoveFromCart(product)}
        />
      </PaperStyle>
    );
  }
}

const mapStateToProps = (state, props) => ({
  quantity: props.product.quantity
});

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({
    changeQuantity,
    deleteFromCart
  }, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(CartElement);
