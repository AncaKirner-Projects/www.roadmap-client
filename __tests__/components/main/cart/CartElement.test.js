import React from 'react';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import CartElement from '../../../../src/components/main/cart/CartElement';
// import ActionTypes from '../../../src/store/actions/actionTypes';

const middlewares = [thunk];
const mockStore = configureMockStore();

describe('<CartElement />', () => {
  const store = mockStore({
    cart: {
      products: [
        { id: 1, prod_name: 'Prod1' }
      ]
    }
  });

  it('renders an element', () => {
    const product = {
      id: 1,
      prod_name: 'Prod1',
      prod_number: 15,
      quantity: 2
    };
    const wrapper = shallow(<CartElement store={store} product={product} />);
    const component = wrapper.dive();

    expect(toJson(component)).toMatchSnapshot();
  });
});
