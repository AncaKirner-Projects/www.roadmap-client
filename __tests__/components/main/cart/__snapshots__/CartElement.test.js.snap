// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartElement /> renders an element 1`] = `
<Styled(WithStyles(Paper))
  className={
    Object {
      "margin": 100,
      "padding": 200,
    }
  }
>
  <WithStyles(Grid)
    container={true}
    spacing={16}
    wrap="nowrap"
  >
    <WithStyles(Grid)
      item={true}
      xs={true}
      zeroMinWidth={true}
    >
      <WithStyles(Typography)
        noWrap={true}
        variant="subtitle1"
      >
        Prod1
      </WithStyles(Typography)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={true}
      zeroMinWidth={true}
    >
      <TextField
        id="outlined-select-currency-native"
        onChange={[Function]}
        required={false}
        select={true}
        value={2}
        variant="outlined"
      >
        <option
          key="1"
          value={1}
        >
          1
        </option>
        <option
          key="2"
          value={2}
        >
          2
        </option>
        <option
          key="3"
          value={3}
        >
          3
        </option>
        <option
          key="4"
          value={4}
        >
          4
        </option>
        <option
          key="5"
          value={5}
        >
          5
        </option>
        <option
          key="6"
          value={6}
        >
          6
        </option>
        <option
          key="7"
          value={7}
        >
          7
        </option>
        <option
          key="8"
          value={8}
        >
          8
        </option>
        <option
          key="9"
          value={9}
        >
          9
        </option>
        <option
          key="10"
          value={10}
        >
          10
        </option>
        <option
          key="11"
          value={11}
        >
          11
        </option>
        <option
          key="12"
          value={12}
        >
          12
        </option>
        <option
          key="13"
          value={13}
        >
          13
        </option>
        <option
          key="14"
          value={14}
        >
          14
        </option>
        <option
          key="15"
          value={15}
        >
          15
        </option>
      </TextField>
    </WithStyles(Grid)>
    <Styled(WithStyles(Grid))
      item={true}
      zeroMinWidth={true}
    >
      <WithStyles(Typography)
        align="left"
        noWrap={true}
        variant="subtitle1"
      >
        NaN
          Â  RON
      </WithStyles(Typography)>
      <WithStyles(Button)
        onClick={[Function]}
        size="small"
      >
        Remove
      </WithStyles(Button)>
    </Styled(WithStyles(Grid))>
  </WithStyles(Grid)>
  <ResponsiveDialog
    handleClose={[Function]}
    handleConfirmClose={[Function]}
    open={false}
  />
</Styled(WithStyles(Paper))>
`;
